layers/encoders.py

```diff
@@ -194,8 +198,14 @@ def beam_decode(self, encoder_out, encoder_mask, get_input_fn, logprob_fn, bos_i
 
         # Length penalty not needed, because all lengths are 1
 
-        # Select top k hyps to survive to the next time step
-        all_probs, symbols = torch.topk(probs, beam_size, dim=-1) # ([bsz, beam], [bsz, beam])
+        if mode == "best":
+            # Select top k hyps to survive to the next time step
+            all_probs, symbols = torch.topk(probs, beam_size, dim=-1) # ([bsz, beam], [bsz, beam])
+        elif mode == "random":
+            symbols = torch.multinomial(torch.exp(probs), beam_size, replacement=True)
```

Comment:
I think it's better to calculate probs from logits rather than taking exp of logprobs here.

Reply:
gp, will fix

Reply:
oh, but what if logprob_fn is something other than log_softmax?

Reply:
hmm why would we need a logprob_fn other than log_softmax?

Reply:
I don't know...but you made `logprob_fn` an argument to `beam_decode`, so it seems like in principle it could be something else...

At any rate, I am working on an update to this PR but haven't gotten it right yet. So please don't merge.